Added docs for schema definitions. (#5)
data-driven-forms/react-form-renderer: README.md

> # React Form Renderer

Reviewer: Could you add a table of contents?

const fields = [{
  component: componentTypes.TEXT_FIELD,
  name: 'Foo'
}, {
  component: componentTypes.TEXT_FIELD,
  name: 'Bar'
  condition: {
    when: 'Foo',
    is: ['Show Bar field', true, 123, 'Or now'],
  }
}]
```

In example abow, field `Bar` will appear when fields `Foo` value is `'Show bar field'`, `true`, `123` of `'Or now'`.

#### Other attributes

Any other attributes will be passed to the component, that matches the `component` identifier.

For examples definition fo select component might look something like this:

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const field = {
  component: componentTypes.SELECT_COMPONENT,
  name: 'color'
  label: 'Choose your favorite color'
  options: [{
    label: 'Red',
    value: 'red'
  }, {
    label: 'Blue',
    value: 'blue',
  }, {
    label: 'Green',
    value: 'sneaky-red',
  }]
}
```

Remember that the components defines the interface. If your label to be image, just pass it image source with isImage flag maybe and handle rendering in the component.

### Field array and Fixed list
TO DO add documentaion here

## <a name="form-fields-mapper"></a> Component mapping

> In the text above, there was mentioned several times, that you can define your own components for rendering. In fact, you have define them because the Form Renderer does not know anything about them. This way, the renderer is universal and can be used with any component library or you custom components. It is also very easy to swap look and feel of form without any changes to the format!

Reviewer: You have to define

Your custom componentes

I would use an article before the look (or use appereance)

Feel -> feeling (but i will probably remove feeling at all)

> We also understand that writing form components from the scratch might not be very user friendly. ManageIQ and Insigins are using [patternfly style patterns](http://patternfly-react.surge.sh/), so you can inspire yourself with [Patternfly](https://github.com/data-driven-forms/pf3-component-mapper) 3 and [Patternfly 4](https://github.com/data-driven-forms/pf4-component-mapper) mappers.

Reviewer: Insights

> Form renderer requires twi different mappers. Layout mapper and Form  FieldsMapper.

Reviewer: Two

  [layoutComponents.BUTTON_GROUP]: 'ButtonGroup',
  [layoutComponents.ICON]: 'Icon',
  [layoutComponents.ARRAY_FIELD_WRAPPER]: 'ArrayFieldWrapper',
  [layoutComponents.HELP_BLOCK]: 'HelpBlock'
}
```

LayoutMapper is just good old javascript object, with keys from layoutComponents, and values are just React components:

#### FormWrapper

Form wrapper is your form wrapper component. Typically it will be your react version of `<form>` tag:

```JSX
import FormRenderer, { layoutComponents } from '@data-driven-forms/react-form-renderer';

const FormWrapper = ({ children }) => <form>{children}</form>

const layoutMapper = {
  [layoutComponents.FORM_WRAPPER]: FormWrapper,
}

const MyForm = () => (
  <FormRenderer
    layoutMapper={layoutMapper}
  />
)
```

#### Button
Button component will be used for your submit, reset and cancel buttons.

```JSX
import layoutComponents } from '@data-driven-forms/react-form-renderer';

const Button = ({ label, bsStyle, ...props }) => (
  <button
    {...props}
    style={{ backgroud: bsStyle === primary ? 'blue' : 'initial' }}
  >
    {label}
  </button>
)
const layoutMapper = {
  [layoutComponents.BUTTON]: FormWrapper,
}
```

#### Col

> Col represents wrapper arround one Form Field (hence the name Col). It does not have to mirror bootstrap Col, that is just the name we have decided to go with. If you for instance don'n need any Col (or ther wrapping) component, and you are handling this inside the actual Field component you can use `<React.Fragment>` component. This way you will not create any emelent in your DOM. On the other hand, it might be usefull as a container for your components. Because we can't posibbly create layout that will suit 100% use cases, use can use this wrapper to pass additional styles to field components.

Reviewer: Don’t
Element
Possibly
100% of
You can use

# React Form Renderer

Rendering forms from data structures

- [ Instalation ](#instalation)
- [ Getting started ](#getting-started)
- [ Form schemas ](#form-schemas)
  - [ Default schema ](#default-schema)
    - [ `Default schema attributes` ](#default-schema-attributes)
    - [ `Fields` ](#fields)
    - [ `Field attributes` ](#field-attributes)
- [ Component mappers ](#component-mappers)
  - [ Layout mapper ](#layout-mapper)
  - [ Form Fields Mapper ](#form-fields-mapper)

## <a name="instalation"></a>Instalation

```
npm install --save @data-driven-forms/react-form-renderer
```
or
```
yarn add @data-driven-forms/react-form-renderer
```

## <a name="getting-started"></a>Getting started

> React form renderer is a component designed for ManageIQ and Insighs projects, that takes json form definitions and renders them into react components. It is using [React final form](https://github.com/final-form/react-final-form) for the form state management. It is highly recommended to check their documentations first, to fully understand how the [data-driven-forms](https://github.com/data-driven-forms) libraries work.

Reviewer 2: bez carky

```javascript
const fields = [{...}, [{...}, {...}], {...}, {...}, [[[{...}]]]]
```

In human language, items of field array must be either objects, where each object represents one formField (React component), or array of objects, which are form fields too. This rule allows the component to render all fields, in one cycle with minimal code branching.

The structure of single object is following:

#### <a name="field-attributes"></a> `field attributes`

There are listed all field (items of the `fields` array) attributes, that are defined by the default schema. Any other attributes given to field object are automatically passed to the specified component.

Detailed descriptions of each attribute is bellow.

|name|type|
|---|---|
|`component`|string|
|`name`| string|
|`validate?`|Array of Objects|
|`condition?`|Object|
|`dataType`|string|
|`assignFieldProvider?`|bool|
|SubForm only||
|`fields`| Array (only for `SUB_FORM`)|
|`title?`|string|
|`description?`|string|

```javascript
import { componentTypes, validatorTypes } from '@data-driven-forms/react-form-renderer';

const field = {
  component: componentTypes.TEXT_FIELD,
  name: 'first-name',
  label: 'First name',
  ...
}
```

> Note that the field structure may vary based on your component implementation. There are a few required attributes and most of the does not have to match given types. Most of them are based on used form components.

Reviewer 2: ```suggestion
Note that the field structure may vary based on your component implementation. There are few required attributes and most of them do not have to match the given types. Most of them are based on used form components.
```

#### `component: string`

> Unique identifier of the component. Final component will be picked based in this key. There are several pre-defined constants identifying the most common components for ManageIQ and Insigths apps.

Reviewer 2: ```suggestion
Unique identifier of the component. Final component will be picked based on this key. There are several pre-defined constants identifying the most common components for ManageIQ and Insights apps.
```

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

componentTypes = {
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  FIXED_LIST: 'fixed-list',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  TAG_CONTROL: 'tag-control',
}
```

> You are not limited by these. You can add your own or use only a few of them or combination of both. More detailed explanation of how this will impact the rendered form [can be found here](#form-fields-mapper).

Reviewer 2: ```suggestion
We are not limited by these (these what? name it, values?). You can add your own (?values?)or use only few of them or combination of both. More detailed explanation of how this impacts the rendered form [can be found here](#form-fields-mapper).
```

#### `name`
This is traditional html5 name attribute for input elements.

#### `label`

> Label for form field. Its type is based on your component definition.

Reviewer 2: ```suggestion
Label for form field. The type is based on your component definition.
```

validatorTypes = {
  REQUIRED: 'required-validator',
  /**
  * min length if the input value
  */
  MIN_LENGTH: 'min-length-validator',
  /**
  * minimum count of fileds in some dynamic list of fields
  */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
  * Minimum value of number input
  */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
   * Maximum value of number inpuy
  */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
  * Regexp pattern validator
  */
  PATTERN_VALIDATOR: 'pattern-validator',
}

const validate = [{
  type: validatorTypes.REQUIRED,
  message: 'This is custom error message'
}, {
  type: validatorTypes.MIN_LENGTH,
  treshold: integer
}, {
  type: validatorTypes.MIN_ITEMS_VALIDATOR,
  treshold: integer
}, {
  type: validatorTypes.MIN_NUMBER_VALUE
  value: integer
}, {
  type: validatorTypes.MAX_NUMBER_VALUE
  value: integer
}, {
  type: PATTERN_VALIDATOR,
  pattern: string // regex pattern
  showPatter: bool? // if message is not define turns on/of pattern in error message
}]
```
Validation functions are triggered only when field has a value with exception of required validator.

#### `dataType: string?`

> In addition to explicitly defined validation, you can also add validation based on data type of the field. This can be helpfull in cases, where you can enforce data type with input type.

Reviewer 2: ```suggestion
In addition to explicitly defined validation, you can also add validation based on data type of the field. This can be helpfull in cases, where you can enforce data type with input type. celou tuhle vetu nechapu, zkus to preformulovat
```

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const field = {
  component: componentTypes.TEXT_FIELD,
  name: 'number',
  type: 'number',
  label: 'Integer number',
  dataType: 'integer',
}
```

There are currently four defined data types:
```javascript
['integer', 'number', 'bool', 'string']
```

#### <a name="assign-field-provider"></a> `assignFieldProvider: bool?`

> FieldProvider is just a fancy name for [Field component](https://github.com/final-form/react-final-form#field--reactcomponenttypefieldprops). Following component types are wrapper by it by default:

Reviewer 2: ```suggestion
FieldProvider is just a fancy name for [Field component](https://github.com/final-form/react-final-form#field--reactcomponenttypefieldprops). Following component types are wrapped in the FieldProvider by default:
```

#### `condition: Object?`
Condition is used to define condition fields. For instance, field **A** should render only when field **B** has value **Foo**.

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const fields = [{
  component: componentTypes.TEXT_FIELD,
  name: 'Foo'
}, {
  component: componentTypes.TEXT_FIELD,
  name: 'Bar'
  condition: {
    when: 'Foo',
    is: 'Show Bar field',
  }

}]
```

Sometimes you might want to show field when it's matching multiple values:

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const fields = [{
  component: componentTypes.TEXT_FIELD,
  name: 'Foo'
}, {
  component: componentTypes.TEXT_FIELD,
  name: 'Bar'
  condition: {
    when: 'Foo',
    is: ['Show Bar field', true, 123, 'Or now'],
  }
}]
```

In example above, field `Bar` will appear when fields `Foo` value is `Show bar field`, `true`, `123` or `Or now`.

#### Other attributes

> Any other attributes will be passed to the component, that matches the `component` identifier.

Reviewer 2: ```suggestion
Any other attributes will be passed to the component matching the `component` identifier.
```

For examples definition of select component might look something like this:

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const field = {
  component: componentTypes.SELECT_COMPONENT,
  name: 'color'
  label: 'Choose your favorite color'
  options: [{
    label: 'Red',
    value: 'red'
  }, {
    label: 'Blue',
    value: 'blue',
  }, {
    label: 'Green',
    value: 'sneaky-red',
  }]
}
```

> Remember that the components define the interface. If your label should be image, just pass it image source with isImage flag maybe and handle rendering in the component.

Reviewer 2: ```suggestion
Remember that the components define the interface. If your label is an image, pass the image source with isImage flag maybe and handle rendering in the component.
```

### Field array and Fixed list
TO DO add documentaion here

> ## <a name="component-mappers"></a> Component mapping

Reviewer 2: ```suggestion
## <a name="component-mappers"></a> Component Mapping
```

> In the text above it was mentioned several times, that you can define your own components for rendering. In fact, you have to define them because the Form Renderer does not know anything about them. This way, the renderer is universal and can be used with any component library or your custom components. It is also very easy to swap look of form without any changes to the format!

Reviewer 2: ```suggestion
As it was already mentioned, you can define your own components for rendering. In fact, you have to define them because the Form Renderer does not know anything about them. This way, the renderer is universal and can be used with any component library or your custom components. It is also very easy to swap the look of the form without any changes to the format!
```

> We also understand that writing form components from the scratch might not be very user friendly. ManageIQ and Insights are using [patternfly style patterns](http://patternfly-react.surge.sh/), so you can inspire yourself with [Patternfly 3](https://github.com/data-driven-forms/pf3-component-mapper) and [Patternfly 4](https://github.com/data-driven-forms/pf4-component-mapper) mappers.

Reviewer 2: ```suggestion
We also understand that writing form components from the scratch might not be very user friendly. ManageIQ and Insights are using [patternfly style patterns](http://patternfly-react.surge.sh/), so you can get inspired with [Patternfly 3](https://github.com/data-driven-forms/pf3-component-mapper) and [Patternfly 4](https://github.com/data-driven-forms/pf4-component-mapper) mappers.
```

Form renderer requires two different mappers. Layout mapper and Form  FieldsMapper.

### <a name="layout-mapper"></a> Layout mapper

> Component inside this mapper, influence the layout of the form. Now compared to the Form Fields mapper, it is stric, because you cant define your own elements. Layout mapper must contain these types components: 

Reviewer 2: ```suggestion
Component inside this mapper influence the layout of the form. Now compared to the Form Fields mapper, we have to be very strict because we cannot define your own elements. Layout mapper must contain these types components: 
```

```javascript
import { layoutComponents } from '@data-driven-forms/react-form-renderer';

const layoutComponents = {
  [layoutComponents.FORM_WRAPPER]: 'FormWrapper',
  [layoutComponents.BUTTON]: 'Button',
  [layoutComponents.COL]: 'Col',
  [layoutComponents.FORM_GROUP]: 'FormGroup',
  [layoutComponents.BUTTON_GROUP]: 'ButtonGroup',
  [layoutComponents.ICON]: 'Icon',
  [layoutComponents.ARRAY_FIELD_WRAPPER]: 'ArrayFieldWrapper',
  [layoutComponents.HELP_BLOCK]: 'HelpBlock'
}
```

> LayoutMapper is just good old javascript object, with keys from layoutComponents, and values are just React components:

Reviewer 2: ```suggestion
LayoutMapper is just good old javascript object with keys from layoutComponents, and the values are just React components:
```

#### FormWrapper

Form wrapper is your form wrapper component. Typically it will be your react version of `<form>` tag:

```JSX
import FormRenderer, { layoutComponents } from '@data-driven-forms/react-form-renderer';

const FormWrapper = ({ children }) => <form>{children}</form>

const layoutMapper = {
  [layoutComponents.FORM_WRAPPER]: FormWrapper,
}

const MyForm = () => (
  <FormRenderer
    layoutMapper={layoutMapper}
  />
)
```

#### Button
Button component will be used for your submit, reset and cancel buttons.

```JSX
import layoutComponents } from '@data-driven-forms/react-form-renderer';

const Button = ({ label, bsStyle, ...props }) => (
  <button
    {...props}
    style={{ backgroud: bsStyle === primary ? 'blue' : 'initial' }}
  >
    {label}
  </button>
)
const layoutMapper = {
  [layoutComponents.BUTTON]: FormWrapper,
}
```

#### Col

> Col represents wrapper arround one Form Field (hence the name Col). It does not have to mirror bootstrap Col, that is just the name we have decided to go with. If you for instance don't need any Col (or other wrapping) component, and you are handling this inside the actual Field component you can use `<React.Fragment>` component. This way you will not create any element in your DOM. On the other hand, it might be usefull as a container for your components. Because we can't posibbly create layout that will suit 100% of use cases, you can use this wrapper to pass additional styles to field components.

Reviewer 2: ```suggestion
Col represents wrapper arround one Form Field (hence the name Col). It does not have to mirror bootstrap Col, which is just the name we have decided to go with. If you for instance don't need any Col (or other wrapping) component, and you are handling this inside the actual Field component, you can use `<React.Fragment>` component. This way you will not create any element in your DOM. On the other hand, it might be usefull ??to have it - chybi sloveso??as a container for your components. Because we can't possibly create layout that  suit 100% of our use cases, we can use this wrapper to pass additional styles to field components.
```

const Col = ({ children }) => (
  <div className="form-row">{children}</div>
)
```

#### FormGroup

Very similar to Col component.

```JSX
import './form/styles.scss';

const FormGroup = ({ children }) => (
  <div className="form-group">{children}</div>
)
```

#### ButtonGroup
Wrapper for your form buttons

```JSX
import './form/styles.scss';

const ButtonGroup = ({ children }) => (
  <div className="button-group">{children}</div>
)
```

#### Icon, Array Field Wrapper, Help Block
TO DO when array field docs are done

#### Putting it all together

```JSX
import FormRenderer, { layoutComponents } from '@data-driven-forms/react-form-renderer';

const layoutMapper = {
  [layoutComponents.FORM_WRAPPER]: FormWrapper,
  [layoutComponents.BUTTON]: Button,
  [layoutComponents.COL]: Col,
  [layoutComponents.FORM_GROUP]: FormGroup,
  [layoutComponents.BUTTON_GROUP]: ButtonGroup,
  [layoutComponents.ICON]: Icon,
  [layoutComponents.ARRAY_FIELD_WRAPPER]: ArrayFieldWrapper,
  [layoutComponents.HELP_BLOCK]: HelpBlock
}

const MyForm = () => (
  <FormRenderer
    ...
    layoutMapper={layoutMapper}
    ...
  />
)
```

### <a name="form-fields-mapper"></a> Form Fields mapper

> Unlike the layout components, form fields are completely customizable, and the implementation is restricted to only one rule. In order to correctly change the form state you have to use provided `input` and `meta` props on your input fields. These objects provide functions like `onChange`, `onBlur`, error messages, valid state and more. Again you should probably read more about then in the [React Final Form docs](https://github.com/final-form/react-final-form#field--reactcomponenttypefieldprops).

Reviewer 2: ```suggestion
Unlike the layout components, the form fields are completely customizable, and the implementation is restricted to only one rule. In order to correctly change the form state, you have to use provided `input` and `meta` props on your input fields. These objects provide functions like `onChange`, `onBlur`, error messages, valid state and more. Again you should probably read more about that in the [React Final Form docs](https://github.com/final-form/react-final-form#field--reactcomponenttypefieldprops).
```

> As said before, some of the component types are wrapped in this component [by default](#assign-field-provider) and if you need to use different component type, you can pass it as a prop by adding the `assignFieldProvider: true` attribute to field.

Reviewer 2: ```suggestion
Some of the component types are wrapped in this (what? component) component [by default](#assign-field-provider) and if you need to use different component type, you can pass it as a prop by adding the `assignFieldProvider: true` attribute to the field.
```

> In example bellow you can see implementation if simple input component using both predefined component type and custom one

Reviewer 2: ```suggestion
In an example below you can see an implementation of a simple input component using both predefined component type and a custom one.
```

```JSX
import { componentTypes, validatorTypes } from '@data-driven-forms/react-form-renderer';

// if no id is provided field will use name and assign it to ID
const DefaultInput = ({ input, meta, label, name, ...rest }) => {
  const id = rest.id || name;
  return (
    <div className={`form-group ${meta.error ? 'error' : ''}`}>
      <label htmlFor={id}>{label}</label>
      <input id={id} name={name} {...rest} {...input} />
      {meta.error && <span className="error-text">{meta.error}</span>}
    </div>
  )
}

const CustomInput = ({ FieldProvider, ...rest }) => (
  <FieldProvider
    {...rest}
    component={DefaultInput}
  />
)

const formFieldsMapper = {
  [componentTypes.TEXT_FIELD]: DefaultInput,
  'custom-input': CustomInput,
}

const schema = {
  fields:[{
    component: componentTypes.TEXT_FIELD.
    name: 'first-name',
    label: 'First Name',
    type: 'text',
    validate: [{
      type: validatorTypes.REQUIRED,
    }]
  }, {
    component: 'custom-input'.
    name: 'last-name',
    label: 'Last Name',
    type: 'text',
    assignFieldProvider: true,
    validate: [{
      type: validatorTypes.REQUIRED,
    }]
  }]
}

```

#### What about nesting?

> You might sometimes need something like a SubForm in your Forms. Fortunatelly this is allso possible. There are two ways to do it.

Reviewer 2: ```suggestion
There might be a need to create something like a SubForm in your Forms. We made it possible, and there are two ways to do it.
```

1. Use pre defined `componentTypes.SUB_FORM` component type:
```JSX
import { componentTypes } from '@data-driven-forms/react-form-renderer';

const schema = {
  fields: [{
    component: componentTypes.TEXT_FIELD,
    name: 'first-name',
    ...
  }, {
    component: componentTypes.SUB_FORM,
    title: 'Address',
    description: 'Please provide us with your shipping address.',
    name: 'address',
    fields: [{
      component: componentTypes.TEXT_FIELD,
      name: 'city',
      label: 'City',
      validate: [...],
      ...
    }]
  }]
}
```

> This will render and bundle fields and their values inside the `SUB_FORM` component. When you submit the form, the values object will look somethink like this:

Reviewer 2: ```suggestion
This renders and bundles fields and their values inside the `SUB_FORM` component. When the form is submitted, the values object will look something like this:
```

```json
{
  "first-name": "Bob",
  "address": {
    "city": "Prague"
    ...
  }
}
```

2. Custom component

> Again, the `SUB_FORM` component type might not be the correct solution for your use case. So you can always create your own. Lets use the same schema as in example above, with one change:

Reviewer 2: ```suggestion
Again, the `SUB_FORM` component type might not be the correct solution for your use case. You can always create your own. Let's use the same schema as in example above, but only with one change:
```

```javascript
component: componentTypes.SUB_FORM, -> component: 'custom-sub-form'
```

> If you use custom type, in addition to all the attributes in the field specification, you will also receive `formOptions` props. Form options contains some global form properties, and most importantly, there is a `renderForm` function. It does exactly what it says. It is the main loop that renders the form. It acceps two arguments, `fields` and `formOptions`:

Reviewer 2: ```suggestion
If you use custom type, in addition to all the attributes in the field specification, you will also receive `formOptions` props. Form options contains some global form properties, and most importantly, there is a `renderForm` function. It does exactly what it says. It is the main loop that renders the form. It accepts two arguments, `fields` and `formOptions`:
```

# React Form Renderer

Rendering forms from data structures

- [ Instalation ](#instalation)
- [ Getting started ](#getting-started)
- [ Form schemas ](#form-schemas)
  - [ Default schema ](#default-schema)
    - [ `Default schema attributes` ](#default-schema-attributes)
    - [ `Fields` ](#fields)
    - [ `Field attributes` ](#field-attributes)
- [ Component mappers ](#component-mappers)
  - [ Layout mapper ](#layout-mapper)
  - [ Form Fields Mapper ](#form-fields-mapper)

## <a name="instalation"></a>Instalation

```
npm install --save @data-driven-forms/react-form-renderer
```
or
```
yarn add @data-driven-forms/react-form-renderer
```

## <a name="getting-started"></a>Getting started

> React form renderer is a component designed for ManageIQ and Insighs projects that takes json form definitions and render them into react components. It uses [React final form](https://github.com/final-form/react-final-form) for the form state management. It is highly recommended to check their documentations first to fully understand how the [data-driven-forms](https://github.com/data-driven-forms) libraries work.

Reviewer 3: render**s** them

Code examples:

```JSX
import FormRender from '@data-driven-forms/react-form-renderer';

const DataDrivenForm = () => (
  <FormRender
    formFieldsMapper={formFieldsMapper}
    layoutMapper={layoutMapper}
    schema={schema}
    onSubmit={...}
    onCancel={...}
    onReset={...}
    canReset
  />
);
```

There are several required props that must be passed to the component. Check the examples below to learn how it works.

> ## <a href="form-schemas"></a> Form schemas

Reviewer 4: href vs name?

Author: good catch

There are currently 3 schema definitions you can use to define your forms. With the intention to provide additional customization in the future. Currently supported schemas are:

- [default schema](#default-schema)
- ManageIQ schema

> - mozilla json schema

Reviewer 3: Mozilla

### <a name="default-schema"></a> Default schema
This is the default schema that is used directly for rendering the form. All other schema types are parsed to this one. This gives the option to write your custom parser that transforms any of your existing definitions into the default one, and use this renderer.

> The default schema is also very extensible. There is only a few requirements for the format. Most of the attributes is meta information and their shape is based upon **your** form components.

Reviewer 3: attributes is --> attributes are

|`title?`|string|
|`description?`|string|

```javascript
import { componentTypes, validatorTypes } from '@data-driven-forms/react-form-renderer';

const field = {
  component: componentTypes.TEXT_FIELD,
  name: 'first-name',
  label: 'First name',
  ...
}
```

Note that the field structure may vary based on your component implementation. There are few required attributes and most of them do not have to match the given types. Most of them are based on used form components.

#### `component: string`
Unique identifier of the component. Final component will be picked based on this key. There are several pre-defined constants identifying the most common components for ManageIQ and Insights apps.

```javascript
import { componentTypes } from '@data-driven-forms/react-form-renderer';

componentTypes = {
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  FIXED_LIST: 'fixed-list',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  TAG_CONTROL: 'tag-control',
}
```

> We are not limited by these component types. You can add your type own or use only few of them or combination of both. More detailed explanation of how this impacts the rendered form [can be found here](#form-fields-mapper).

Reviewer 3: " can add your type own " ---> can add your own type

#### `name: string`
This is traditional html5 name attribute for input elements.

#### `label`
Label for form field. The type is based on your component definition.

#### `validate: Array?<Object>`
Array of validation definitions. These are limited by the form renderer (Might be configurable in future).

> If you want to use out of the box validation, you must use this defined format:

Reviewer 3: use this defined format --> use this format
